### PR TITLE
Search

### PR DESCRIPTION
**Pull request by genuineAB**
## Changes Made

- search controller, gateway, route, schema, and usecase were created
- config, entities, and interface were modified 

## How to test

- [Search API](https://github.com/users/9jacodeTech/projects/2/views/4?pane=issue&itemId=39049268)
-
-

## **Screenshots**
image

### CODE REVIEW

File: src/data-gateway/search-data-gateway.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+];
+
+export class SearchDataGateway implements ISearchDataGateway {
+  constructor(private readonly fileService: IFileService) {}
+
+  async fetch(data: SearchDetailInput): Promise<any[]> {
+    const fileNameCategory = this.getFileNameForCategory(data.category);
+
+    try {
+      var results: any[] = [];
```
Comment (chidioguejiofor) [inline src/data-gateway/search-data-gateway.ts]: dont use var please

Comment (chidioguejiofor) [inline src/data-gateway/search-data-gateway.ts]: The logic in this should actually be in the search usecase. 

Here is the idea. The search usecase should have all these other gateways as its dependencies(ie the constructor).

File: src/data-gateway/search-data-gateway.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+  private getFileNameForCategory(category: string): string {
+    const categoryToMap = {
+      news: [NEWS_FILE_ENV],
+      events: [EVENTS_FILE_ENV],
+      projects: [PROJECT_FILE_ENV],
+      programs: [PROGRAM_FILE_ENV],
+    };
+
+    return categoryToMap[category] || fileName;
```
Comment (chidioguejiofor) [inline src/data-gateway/search-data-gateway.ts]: The logic of this search needs to not know where the data is being stored. 

In other words, we should depend on the data gateways to get these values

File: src/usecases/search.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { type SearchDetailInput } from 'entities';
+import { type ISearchDataGateway } from './interfaces';
+import { validateData } from 'utils/helpers';
+import { searchInputSchema } from 'schemas/search';
+
+export class SearchUsecase {
+  constructor(private readonly dataGateway: ISearchDataGateway) {}
```
Comment (chidioguejiofor) [inline src/usecases/search.ts]: This usecase is where all the logic of the search should be. 

What this means is that we add `eventsDatagateway`, `newsDatagateway` etc.

Then we add the logic of the search based on the results we get from the gateways. 

That way one can understand the logic from a high level

**Review by chidioguejiofor — CHANGES_REQUESTED** [review]

File: src/schemas/projects.ts [hunk 4]
```diff
@@ -3,10 +3,19 @@ import Joi from 'joi';
 export const projectInputSchema = Joi.object({
   name: Joi.string().required(),
   date: Joi.date().required(),
-  description: Joi.string().required(),
+  sections: Joi.array()
+    .items(
+      Joi.object({
+        title: Joi.string().optional(),
+        body: Joi.string().optional(),
+        imageUrl: Joi.array().items(Joi.string()).optional(),
```
Comment (chidioguejiofor) [inline src/schemas/projects.ts]: imageURL

Comment (genuineAB) [reply src/schemas/projects.ts]: I have made the necessary changes requested

**Pull request merged**